src/rule_dsl.jl

```diff
@@ -248,19 +248,41 @@ struct RuleRewriteError
     expr
 end
 
-function (r::RuleSet)(term, context=EmptyCtx(); depth=typemax(Int), applyall=false, recurse=true)
+node_count(atom, count; cutoff) = count + 1
+node_count(t::Term, count=0; cutoff=100) = sum(node_count(arg, count; cutoff=cutoff) for arg ∈ arguments(t))
+
+function _recurse_apply_ruleset_threaded(r::RuleSet, term, context; depth, thread_subtree_cutoff)
+    _args = map(arguments(term)) do arg
+        if node_count(arg) > thread_subtree_cutoff
+            Threads.@spawn r(arg, context; depth=depth-1, threaded=true,
+                             thread_subtree_cutoff=thread_subtree_cutoff)
+        else
+            r(arg, context; depth=depth-1, threaded=false)
+        end
+    end
+    args = map(t -> t isa Task ? fetch(t) : t, _args)
+    Term{symtype(term)}(operation(term), args)
+end
+
+function (r::RuleSet)(term, context=EmptyCtx();  depth=typemax(Int), applyall::Bool=false, recurse::Bool=true,
+                      threaded::Bool=false, thread_subtree_cutoff::Int=100)
```

Comment:
You can probably do the `num_counts` check first thing in this function and call `@spawn r(with the exact same args but threaded=false)` and you would get the same behavior as you have here, but with way less code (not sure if you'll still need that function barrier.)

Reply:
My thinking here was that suppose you have something like
```
term(f, x, some_giant_term, y, a_big_term)
```
you do not want to `@spawn on `r(x)` or  `r(y)`, you only want to spawn on `r(some_giant_term)` and `r(a_big_term)`. 

The function barrier is because `@spawn` generates a large amount of code and can cause the optimizer to give up inside the ruleset application, so moving it outside helps things.